doc/read-the-docs-site/adr/0005-plutus-tx-compiler-and-library.md

```diff
@@ -0,0 +1,205 @@
+
+```scala
+def foo(a: Int, b: => Int) { ... }
+```
+
+`nonstrict` can be used to achieve a similar effect, except that `nonstrict` is used at the call sites as opposed to function definitions.
+
+#### `foldr` and `nonstrict` in Other Strict Languages
+
+Scala and OCaml are both strict by default.
+In both languages, their version of `foldr` for lists in the respective standard library ([Scala](https://www.scala-lang.org/api/2.13.3/scala/collection/immutable/List.html#foldRight[B](z:B)(op:(A,B)=%3EB):B),[OCaml](https://v2.ocaml.org/api/List.html)) is strict and doesn't short circuit.
+Functions that should short-circuit, such as `all` and `any`, do short-circuit, and unsurprisingly they are not implemented in terms of `foldr`.
+
+Scala's [cats library](https://github.com/typelevel/cats) provides a `Foldable` typeclass, and a short-circuiting `foldRight`.
+In order to do so, cats' `foldRight` has a different type than the one in the standard library: the type of the accumulator is `Eval b` rather than `b`.
+Our `nonstrict` function is similar to `Eval.always`, except that `nonstrict`'s type is `a -> a` while the type of `Eval.always` is `a -> Eval a` - because it is a library, not a plugin.
+Incidentally, there's a simpler way to make `foldRight` lazy using Scala's built-in syntax for lazy arguments, but it is not stack safe.
+Interestingly, the `Foldable.forall` and `Foldable.exists` functions in cats are implemented by [converting the foldable structure to list first](https://github.com/typelevel/cats/blob/896f00cd689af20676556d1b57791d7ade5cfa04/core/src/main/scala/cats/Foldable.scala#L677-L686), rather than using `foldRight`.
+
+We can potentially introduce something similar to `Eval` in the Plutus Tx standard library.
+The advantage of `Eval` is that it can be provided entirely in a library, without special compiler support. However, this is also what leads to its disadvantage: `Eval` is more heavyweight to use, since it changes the types of many functions (such as `foldr`), while `nonstrict` has the same type as the identity function, which is possible because it is a magical function that the compiler treats specially.
+
+The takeaway is this: the standard libraries usually do the simple things, e.g., providing a monomorphic, non-short-circuiting version of `foldr` for lists, as well as functions like `all`, `any`, also monomorphically for lists.
+The polymorphic versions of those functions, if exist, are usually provided by third-party libraries like cats.
+
+For Plutus Tx, even with `nonstrict`, it is still useful to provide monomorphic versions of certain functions in the standard library, especially since they can be noticeably cheaper than the polymorphic versions.
+This will be discussed in the next section.
+
+### Monomorphizing Plutus Tx Standard Library
+
+Strict languages do not lend themselves to code reuse, as there is a high cost to use abstractions such as higher-order functions and parametric polymorphism.
```

Comment:
Why is parametric polymorphism expensive? Do you mean ad-hoc polymorphism (i.e. typeclasses)? If so that's not really about strict languages: GHC would have exactly the same overheads as we do (and often does!), but also goes to some lengths to eliminate dictionary passing.

Reply:
Yes, I meant ad-hoc polymorphism. I think it is more of a problem for strict languages. The example I had in mind was: it is difficult to implement a short-circuiting `any :: Foldable t => (a -> Bool) -> t a -> Bool`. So the cost of ad-hoc polymorphism in this case is the lack of short-circuiting.

I agree the dictionary thing is not unique to strict languages.

Reply:
Short-circuting is, in general, a pretty specific problem. Most uses of ad-hoc polymorphism don't particularly have anything to do with short-circuiting. It's pretty much only folding over lists with boolean operations where people expect this. So I think the framing is too harsh (and also I didn't get what it was saying).

Reply:
See the updated draft, where this statement is removed.